### PR TITLE
Updates to TRAKHTENBROT for the LMCS submission

### PR DESCRIPTION
**Pull request by DmxLarchey**
Dear @yforster and @mrhaandi (sorry for the lengthy description),

This PR is designed to integrate some modifications that were made to the `TRAKHTENBROT` development in sync with the submission of an extended version for the selected papers of IJCAR 2020 paper, submitted to LMCS via Arxiv:

[_Trakhtenbrot's Theorem in Coq: Finite Model Theory through the Constructive Lens_](https://arxiv.org/abs/2104.14445) by  @dominik-kirst and @DmxLarchey 

On top of some restructuring, the paper improves on the IJCAR paper with a section developing links with _Separation logics_ (already part of `coq-8.13`) and another section with a finer characterisation of the requirement for decidable finite satisfiability (FSAT) over monadic signatures, centered on the notion of _(computable) discernability_. So this PR integrate code changes for these improvements of results on decidability, described below. 

We are aware that the `TRAKHTENBROT` directory is _not yet_ in conformity with the guidelines for structuring sub-directories. However, we had to keep it in sync with the submission and the corresponding [`coqdoc`](https://www.ps.uni-saarland.de/extras/fol-trakh-ext/website/Undecidability.TRAKHTENBROT.summary.html). Our intent is to publish a _tag_ and a _release_ named `TRAK-LMCS` which would slightly upgrade on this PR:

+ with a tailored [`Readme.md`](https://github.com/DmxLarchey/coq-library-undecidability/blob/improv_trak_8.13/README.md) file containing specific instructions for compiling just the code relevant to the paper;
+ with a [`TRAKHTENBROT/summary.v`](https://github.com/DmxLarchey/coq-library-undecidability/blob/improv_trak_8.13/theories/TRAKHTENBROT/summary.v) linking the paper results with the definitions/theorems names in the code;
+ these two files are already prepared but they do not fit in `coq-8.13`;
+ then @dominik-kirst will modify the link in http://www.ps.uni-saarland.de/extras/fol-trakh-ext/ to point to the release; right now it points to branch [`improv_trak_8.13`](https://github.com/DmxLarchey/coq-library-undecidability/tree/improv_trak_8.13) in @DmxLarchey's fork of the project.

After that release, we will commit ourselves to restructure the `TRAKHTENBROT` directory to more closely follow the guidelines. and publish a new PR ASAP. The question of whether companion decidability results fit in this library over the long run is open for discussions. From @DmxLarchey's point of view, it could fit since with get some kind of classification for FSAT depending on the properties of FO signatures.

Now some details about this PR. First some various improvements over the previous code:

+ Improved notations to correspond more closely to the paper (LMCS)
+ a new finite quotient along a partial equivalence relation and a list of defined elements (not necessarily all of them) in `Shared/Libs/DLW/Utils/fin_quotient.v`
+ `list_seteq` in `Shared/Libs/DLW/Utils/seteq.v`:
  - a nice inductive characterization of extensional identity of lists (same members)
  - `list_seteq` is permutation augmented with contraction equivalence
  - used to simplify the proof of `bti_equiv` in `btree.v`.
+ better automation in `btree.v`, most proofs are now very short.
+ cleanup (removed dead code) and better notations in `membership.v`
+ simplified the reduction from `Σn` to `Σ2` by removing some unnecessary axioms, impacts `reln_hfs.v` and `Sign_Sig2.v`
+ a better tactic notation `fol equiv *` for proving `A o B <-> A' o B'` where `o` is a binary connective or other. Not all the code is fully updated to use this tactic. 

A finer characterization for decidability of `FSAT` over monadic signatures (arity less than one):

+ as in the `discrete` signature case, one has to distinguish if there is a unary relation or else if all relation symbols are nullary;
+ two symbols are _discernable_ if they can be mapped to two different values in `bool`, eg `f x = true` and `f y = false` for some `f : rels -> bool` (equivalently, they can be mapped to two different values of a discrete type);
+ if there is a unary relation symbol: _`FSAT` is decidable iff discernability of function symbols and of relation symbols are decidable_;
+ if all relation symbols are nullary (arity 0):  _`FSAT` is decidable iff discernability of relation symbols is decidable_.

### CODE REVIEW

**Comment by dominik-kirst** [comment]
> After that release, we will commit ourselves to restructure the `TRAKHTENBROT` directory to more closely follow the guidelines. and publish a new PR ASAP.

Is this work actually still necessary? I've added the interface to the `FOL` folder, so the top-level undecidability statements are already available in the usual format.

> The question of whether companion decidability results fit in this library over the long run is open for discussions.

I'm in favour of allowing such related results, at least if they don't obscure the undecidability results. In fact, I'd even be fine with summary files if they are clearly marked as such and if they are not too exposed.

**Comment by yforster** [comment]
> > After that release, we will commit ourselves to restructure the `TRAKHTENBROT` directory to more closely follow the guidelines. and publish a new PR ASAP.
> 
> Is this work actually still necessary? I've added the interface to the `FOL` folder, so the top-level undecidability statements are already available in the usual format.

I tried to find them, but with no success :) So I'd argue that yes, it would still be good if the `TRAKTHENBROT` directory would contain a specification file and an undecidability file - or if it would disappear altogether and become a subdirectory of `FOL`.
> 
> > The question of whether companion decidability results fit in this library over the long run is open for discussions.
> 
> I'm in favour of allowing such related results, at least if they don't obscure the undecidability results. In fact, I'd even be fine with summary files if they are clearly marked as such and if they are not too exposed.

We already have such decidability results to get a sharp bound in the library, namely in the case of higher-order unification where we show that the first-order case is decidable, and the second-order case with one binary symbol is not (monadic second-order unification is also decidable, but we don't have this tightest possible bound mechanised).

So FTR I'm still in favour of such decidability results as long as they are related to undecidability results.

I also think it's good to have no summary files - it is a pain already to port the library to new versions of Coq. More files won't make this easier. In addition, we might want to refactor some parts of the library further down the road, and summary files make refactorings harder.

The methods of releases / tags / forks of the library for papers work very well from my perspective, so no reason to change here.

File: theories/_CoqProject [hunk 1]
```diff
@@ -1,7 +1,7 @@
 -Q . Undecidability
 
 -arg -w -arg -notation-overridden
-COQDOCFLAGS = "--charset utf-8 -s --with-header ../website/resources/header.html --with-footer ../website/resources/footer.html --index indexpage"
+COQDOCFLAGS = "--charset utf-8 -s --with-header ../website/resources/header.html --with-footer ../website/resources/footer.html --index indexpage --parse-comments"
```
[the remark this answers is not in the document]
Comment (yforster) [reply theories/_CoqProject:4]: Maybe @fakusb also has preferences here he might want to contribute? @mrhaandi is already tagged as a reviewer.

[the remark this answers is not in the document]
Comment (DmxLarchey) [reply theories/_CoqProject:4]: This change in `_CoqProject` was not intended. Only removal of `summary.v` was intended. I will revert it to the original state. Thx for the observation. I will also ask @fakusb.

[the remark this answers is not in the document]
Comment (DmxLarchey) [reply theories/_CoqProject:4]: Oh maybe @dominik-kirst introduced it to generate the CoqDoc. Does it impact anything besides `coqdoc`?

[the remark this answers is not in the document]
Comment (yforster) [reply theories/_CoqProject]: No, it only affects `coqdoc`. The difference is exactly comments being printed in the `html files vs only `(**` comments being printed.

**Review by yforster — APPROVED** [review]

**Review by mrhaandi — APPROVED** [review]

File: theories/_CoqProject (hunk 1 above)
[the remark this answers is not in the document]
Comment (dominik-kirst) [reply theories/_CoqProject]: > Oh maybe @dominik-kirst introduced it to generate the CoqDoc. Does it impact anything besides `coqdoc`?

Yes this was me, I forgot to mention that it's one of the changes we need to roll back for the library PR.

**Pull request merged**